Firestore/core/src/firebase/firestore/local/local_serializer.cc

```diff
@@ -208,6 +211,123 @@ std::unique_ptr<model::NoDocument> LocalSerializer::DecodeNoDocument(
+  if (query.IsDocumentQuery()) {
+    // TODO(rsgowman): Implement. Probably like this (once EncodeDocumentsTarget
+    // exists):
+    /*
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_documents_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeDocumentsTarget(writer, query);
+    });
+    */
+    abort();
+  } else {
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_query_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeQueryTarget(writer, query);
+    });
+  }
+}
+
+util::StatusOr<QueryData> LocalSerializer::DecodeQueryData(
+    const uint8_t* bytes, size_t length) const {
+  Reader reader = Reader::Wrap(bytes, length);
+  QueryData query_data = DecodeQueryData(&reader);
+  if (reader.status().ok()) {
+    return query_data;
+  } else {
+    return reader.status();
+  }
+}
+
+QueryData LocalSerializer::DecodeQueryData(Reader* reader) const {
+  if (!reader->status().ok()) return QueryData::Invalid();
+
+  int64_t target_id;
+  SnapshotVersion version = SnapshotVersion::None();
+  std::vector<uint8_t> resume_token;
+  Query query = Query::Invalid();
+
+  while (reader->bytes_left()) {
+    Tag tag = reader->ReadTag();
+    if (!reader->status().ok()) return QueryData::Invalid();
+
+    switch (tag.field_number) {
+      case firestore_client_Target_target_id_tag:
+        if (!reader->RequireWireType(PB_WT_VARINT, tag))
+          return QueryData::Invalid();
+        target_id = reader->ReadInteger();
+        break;
+
+      case firestore_client_Target_snapshot_version_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        version = SnapshotVersion{reader->ReadNestedMessage<Timestamp>(
+            rpc_serializer_.DecodeTimestamp)};
+        break;
+
+      case firestore_client_Target_resume_token_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        resume_token = reader->ReadBytes();
+        break;
+
+      case firestore_client_Target_query_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        // TODO(rsgowman): Clear 'documents' field (since query and documents
+        // are part of a 'oneof').
+        query =
+            reader->ReadNestedMessage<Query>(rpc_serializer_.DecodeQueryTarget);
+        break;
+
+      case firestore_client_Target_documents_tag:
+        if (!reader->RequireWireType(PB_WT_STRING, tag))
+          return QueryData::Invalid();
+        // Clear 'query' field (since query and documents are part of a 'oneof')
+        query = Query::Invalid();
+        // TODO(rsgowman): Implement.
+        abort();
+
+      default:
+        // Unknown tag. According to the proto spec, we need to ignore these.
+        reader->SkipField(tag);
+        break;
+    }
+  }
+
+  return QueryData(std::move(query), target_id, QueryPurpose::kListen,
```

Comment:
A not specific-to-this-CL question on decoder. Is there a decision what to do with those partial protos sent from server? Say, the proto is field_number_tag,123,snapshot_version_tag,abc,... In abnormal case, the proto we received from server is broken like field_number_tag,abc,... or field_number_rag,123,abc,... then the decoder returns invalid query-data. But if somehow the proto is broken in a somewhat meaningful way,like snapshot_version_tag,abc,... i.e. an entire field is missing, do we want to return a valid query-data which may not be semantically valid? Or there will be another validating code on the result of but not inside of decoder?